botocore/signers.py

```diff
@@ -196,6 +208,65 @@ def generate_presigned_url(self, request_dict, expires_in=3600,
         return request.url
 
 
+class CloudFrontSigner(object):
+
+    @staticmethod
+    def b64encode(content):
+        return base64.b64encode(content).translate(maketrans(b'+=/', b'-_~'))
+
+    def __init__(self, key_id, private_key):
+        """Create a CloudFrontSigner.
+
+        :type key_id: str
+        :param key_id: The CloudFront Key Pair ID
+
+        :type private_key: str
+        :param private_key: The private key string used for signing.
+        """
+        self.key_id = key_id
+        self.rsa_pk = rsa.PrivateKey.load_pkcs1(private_key.encode('utf8'))
+
+    def sign(self, url, expires, starts=None, ip_address=None):
+        """Sign a given url.
+
+        :type url: str
+        :param url: The URL of the protected object.
+
+        :type expires: int
+        :param expires: The expiry time of the URL. Format is a timestamp.
+
+        :type starts: int
+        :param starts: The start time of the URL. Format is a timestamp.
+
+        :type ip_address: str
+        :param ip_address: Use 'x.x.x.x' for an IP, or 'x.x.x.x/x' for a subnet
```

Comment:
Let me chime in on this one as I was not super familiar with cloudfront signing before I looked at the implementation. I also think we should just only allow for the custom/canned policy in the signature and then build abstractions like policy building on top of the signer. Here are the reasons why I think it is appropriate design choice:

1) Sure there are only 3 parameter combinations that cloudfront allows currently, but what stops them from adding more? Everytime they add a new parameter, we will have to go back and update the signer class because it involves the building of the custom policy, and we should not need to because that class's job is strictly to generate urls. This causes to be really a maintenance burden, and has been a pain for us in the past.

2) Botocore is the low-level library. The underlying classes and methods like the signers are not about being super user friendly. They are about being very robust, constant, and reliable. This allows us to build abstractions on top of these classes in other libraries like the CLI and boto3 that are much more user friendly (such as building the policy for the user).

I would recommend taking any of the policy building and move it to a CloudFrontPolicyBuilder class that exists in another module in botocore or lives in a library like the CLI, boto3, or a even new package and leave all of the signing logic in your signer class.